Add GoogleCalendarEvent class with "attendees" attribute

sfdye opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
	- https://community.home-assistant.io/t/google-calendar-invitation-acceptance-status-as-attribute/429269
- Link to documentation pull request: 
	- [ ] Update https://www.home-assistant.io/integrations/google/


Google Calendar [API doc](https://developers.google.com/calendar/api/v3/reference/events#resource)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

sfdye committed:
Add initial code

sfdye commented:
Still WIP, will add the `attendees_omitted` related logic and unit tests later

allenporter reviewed (commented):
Thank you for your contribution.

I am supportive of adding attendees given this is part of rfc5545.  However (1) attendees_omitted is not part of that and (2) a flat list of dictionary attributes is likely not what we want here.

Let's give more visibility to this and open an arch discussion similar to 
https://github.com/home-assistant/architecture/discussions/884
and in that discuss:
(1) the end user use case and some motivating examples of how this will be used to ensure we have the right format (be specific)
(2) the integrations that we plan to make support this. just google calendar probably isn't enough

sfdye commented:
@allenporter Thanks for your input. 

Here we go
https://github.com/home-assistant/architecture/discussions/1099

sfdye committed:
Return accepted events by default

sfdye committed:
Move attendees attribute to new subclass GoogleCalendarEvent

sfdye committed:
WIP: verify attendees[].responseStatus for self

sfdye commented:
Hey @allenporter I am getting back on this (after our discussion [here](https://github.com/home-assistant/architecture/discussions/1099#discussioncomment-9761084)) and making good progress.

Highlight of the new approach:
- Introduce a **new class** `GoogleCalendarEvent` (extend from `CalendarEvent`) and add `attendees` attribute there **only**. This is to control the blast radius and not to worry about other integration use cases just yet.
- Modify the `_event_filter` method to return events that are (both)
	1. Opaque (show as busy)
	2. Accepted by self

I was able to test the changes locally with a few caveats:

1. The logic to filter events that are `accepted by self` needs some love as I wasn't able to figure out where to get the email for self.
2. From my local testing, the returned attendees data seem to be corrupted where `display_name` and `response_status` **always** returns the default value, regardless of the actual value.

Example (real data with redacted email)
```
attendees=[Attendee(id=None, email='redacted', display_name=None, optional=False, comment=None, response_status=<ResponseStatus.NEEDS_ACTION: 'needsAction'>), Attendee(id=None, email='redacted', display_name=None, optional=False, comment=None, response_status=<ResponseStatus.NEEDS_ACTION: 'needsAction'>)]
```

Digging around, I think the issue might be due to the slight nuance of the `alias` feature in pydantic [v1](https://docs.pydantic.dev/1.10/usage/schema/) and [v2](https://docs.pydantic.dev/2.7/api/aliases/). (apparently home assistant is using v1 whereas `gcal_sync` supports both v1 and v2) 
3. The [attendees[].self](https://developers.google.com/calendar/api/v3/reference/events#resource) boolean field might be useful to solve point 1 but it was removed by the [_remove_self](https://github.com/sfdye/gcal_sync/blob/5dcd9da4e94678bece0f220ef61730d798c4c860/gcal_sync/model.py#L92) validator. Can I suggest we keep this and rename it to say `is_self`?

sfdye commented:
Gentle bump

allenporter commented:
Ah sorry you can mark "Ready for review" when its ready, otherwise its still in draft. I'll take a look.

allenporter commented:
sorry didn't realize that was a question above.

I think an upstream to fix `attendees[].self` looks like it may be the way to go. This may need a fix in `gcal_sync` to store it with some other field name since `self` has special meaning in python and pydantic is hiding it. 

I don't think you need to add `GoogleCalendarEvent` given that the filtering is happening on the `gcal_sync` `Event` type.

sfdye commented:
@allenporter 

Thanks for your review on the `attendees[].self` and `GoogleCalendarEvent`. However the whole flow still doesn't work because `attendees[].response_status` is essentially broken due to the [broken alias](https://github.com/home-assistant/core/pull/119457#issuecomment-2183632091) I mentioned above. Do you have any recommendation to fix this? I still pretty new to pydantic.

sfdye commented:
I have created an issue in https://github.com/allenporter/gcal_sync/issues/545

sfdye commented:
> I don't think you need to add GoogleCalendarEvent given that the filtering is happening on the gcal_sync Event type.
But I only want google calendar event to have the `attendees` attribute (i.e. not other local calendar or todoist)

allenporter commented:
> > I don't think you need to add GoogleCalendarEvent given that the filtering is happening on the gcal_sync Event type.
> 
> But I only want google calendar event to have the `attendees` attribute (i.e. not other local calendar or todoist)

The filter function uses `gcal_sync.model.Event` not the calendar platform.

sfdye commented:
> > > I don't think you need to add GoogleCalendarEvent given that the filtering is happening on the gcal_sync Event type.
> 
> > 
> 
> > But I only want google calendar event to have the `attendees` attribute (i.e. not other local calendar or todoist)
> 
> 
> 
> The filter function uses `gcal_sync.model.Event` not the calendar platform.

Right, we don't need to store the attribute, just use it for filtering.

sfdye committed:
Fix event filtering

- Remove GoogleCalendarEvent since the filtering is done on
  gcal_sync.models.Event
- Filter events by using "q" parameter (attendee email)

frenck commented:
Marking this PR draft, awaiting the results of the architectural discussion.

../Frenck
